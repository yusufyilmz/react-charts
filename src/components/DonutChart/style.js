
import styled from 'styled-components';

export const Text = styled.text`
    text-anchor: middle;
    font-weight: bold;
    font-size: larger;
    fill: lightgray;
`;

Text.displayName = 'text'

export const ValueText = styled(Text)`
    font-size: larger;
    fill: black;
    font-size: x-large;
`;


ValueText.displayName = 'text'

