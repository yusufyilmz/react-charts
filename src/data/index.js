
export default [
    {
        title: 'REVENUE',
        totalValue: 200000,
        valueType: 'euro',
        values: [
            {
                name: 'Tablet',
                percentage: 60,
                color: '#47d147'
            },
            {
                name: 'Smartphone',
                percentage: 40,
                color: '#006600'
            }
        ],
        datas: [
            { id: 1, value: 1.1 },
            { id: 2, value: 2.2 },
            { id: 3, value: 1.8 },
            { id: 4, value: 3.2 },
            { id: 5, value: 2.5 },
            { id: 6, value: 5.2 },
            { id: 7, value: 4.6 },
            { id: 8, value: 6.3 },
            { id: 9, value: 4.8 },
        ]
    },
    {
        title: 'IMPRESSIONS',
        totalValue: 200000000,
        valueType: '',
        values: [
            {
                name: 'Tablet',
                percentage: 40,
                color: '#1ad1ff'
            },
            {
                name: 'Smartphone',
                percentage: 60,
                color: '#002699'
            }
        ],
        datas: [
            { id: 1, value: 5.2 },
            { id: 2, value: 4.8 },
            { id: 3, value: 5.0 },
            { id: 4, value: 3.8 },
            { id: 5, value: 2.3 },
            { id: 6, value: 2.8 },
            { id: 7, value: 1.5 },
            { id: 8, value: 2.5 },
            { id: 9, value: 2.0 },
        ]
    },
    {
        title: 'VISITS',
        totalValue: 600000000,
        valueType: '',
        values: [
            {
                name: 'Tablet',
                percentage: 80,
                color: '#ffcc00'
            },
            {
                name: 'Smartphone',
                percentage: 20,
                color: '#e68a00'
            }
        ],
        datas: [
            { id: 1, value: 1.1 },
            { id: 2, value: 2.2 },
            { id: 3, value: 1.8 },
            { id: 4, value: 3.2 },
            { id: 5, value: 2.5 },
            { id: 6, value: 5.2 },
            { id: 7, value: 4.6 },
            { id: 8, value: 6.3 },
            { id: 9, value: 4.8 },
        ]
    },
]

