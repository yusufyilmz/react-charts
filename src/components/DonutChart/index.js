import * as React from "react";
import * as d3 from "d3";
import LineChart from '../LineChart';
import { Text, ValueText } from './style';
import uniqueId from 'lodash.uniqueid';
import ChartPath from '../ChartPath';

const Slice = props => {
    let { pie, colors } = props;

    let interpolateInner = d3.interpolateRgb(colors[0], colors[1]);
    return pie.map((slice, index) => {
        let sliceColor = interpolateInner(index / (pie.length - 1));
        return (
            <ChartPath
                key={uniqueId()}
                {...props}
                radius={130}
                data={slice}
                color={sliceColor} />
        );
    });
};

const DonutChart = ({ values, datas, title, value, colors, width, height }) => {


    let pie = d3.pie()(values);

    return (
        <div>
            <svg height={height} width={width}>
                <g transform={`translate(${width / 2},${height / 2})`}>
                    <Slice
                        key={uniqueId()}
                        pie={pie}
                        colors={colors} />
                    <Text
                        key={uniqueId()}
                        y="-1em">
                        {title}
                    </Text>
                    <ValueText
                        key={uniqueId()}
                        y="0.6em">
                        {value}
                    </ValueText>
                    <LineChart
                        data={datas}
                        width={400}
                        height={200}
                        margin={1}
                        color={colors[0]}
                    />
                </g>
            </svg>
        </div>
    );
};

export default DonutChart;

