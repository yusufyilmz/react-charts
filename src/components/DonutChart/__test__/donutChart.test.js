

import React from 'react';
import DonutChart from '../index';
import { shallow } from 'enzyme'
import data from '../../../data';

describe('DONUTCHART COMPONENT', () => {

    let props, wrapper;

    beforeEach(() => {
        let item = data[0]
        props = {
            height: 400,
            width: 300,
            title: item.title,
            colors: item.values.sort().map(data => data.color),
            value: '',
            values: item.values.sort().map(data => data.percentage),
            datas: item.datas
        }

        wrapper = shallow(
            <DonutChart {...props} />
        );
    });

    test('renders element correctly', () => {
        expect(wrapper.find('text').length).toBe(2)

    });

    test('renders element correctly', () => {
        expect(wrapper.find('svg').length).toBe(1)
    });

    test('renders text correctly', () => {
        expect(wrapper.find('text').at(0).text()).toBe(props.title)
    });

    test('renders text correctly', () => {
        expect(wrapper.find('text').at(1).text()).toBe(props.value)
    });

    test('renders prop correctly', () => {
        expect(wrapper.find('svg').props()['width']).toBe(props.width)
    });
    
    test('renders prop correctly', () => {
        var sty = `translate(${props.width / 2},${props.height / 2})`
        expect(wrapper.find('g').props()['transform']).toBe(sty)
    });

    // test('snapchat correctly', () => {
    //     expect(wrapper).toMatchSnapshot();
    // });

})
