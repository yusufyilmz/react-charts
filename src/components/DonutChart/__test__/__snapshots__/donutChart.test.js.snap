// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DONUTCHART COMPONENT snapchat correctly 1`] = `
<div>
  <svg
    height={400}
    width={300}
  >
    <g
      transform="translate(150,200)"
    >
      <Slice
        colors={
          Array [
            "#47d147",
            "#006600",
          ]
        }
        key="19"
        pie={
          Array [
            Object {
              "data": 60,
              "endAngle": 3.769911184307752,
              "index": 0,
              "padAngle": 0,
              "startAngle": 0,
              "value": 60,
            },
            Object {
              "data": 40,
              "endAngle": 6.283185307179586,
              "index": 1,
              "padAngle": 0,
              "startAngle": 3.769911184307752,
              "value": 40,
            },
          ]
        }
      />
      <text
        key="20"
        y="-1em"
      >
        REVENUE
      </text>
      <text
        key="21"
        y="0.6em"
      />
      <LineChart
        color="#47d147"
        data={
          Array [
            Object {
              "id": 1,
              "value": 1.1,
            },
            Object {
              "id": 2,
              "value": 2.2,
            },
            Object {
              "id": 3,
              "value": 1.8,
            },
            Object {
              "id": 4,
              "value": 3.2,
            },
            Object {
              "id": 5,
              "value": 2.5,
            },
            Object {
              "id": 6,
              "value": 5.2,
            },
            Object {
              "id": 7,
              "value": 4.6,
            },
            Object {
              "id": 8,
              "value": 6.3,
            },
            Object {
              "id": 9,
              "value": 4.8,
            },
          ]
        }
        height={200}
        margin={1}
        width={400}
      />
    </g>
  </svg>
</div>
`;
